Lombiq.Tests.UI/Services/WebDriverFactory.cs

```diff
@@ -18,28 +18,47 @@ public static class WebDriverFactory
+    {
+        var state = new ChromeConfiguration { Options = new ChromeOptions().SetCommonOptions(), Service = null };
 
-                options.AddArgument("--lang=" + configuration.AcceptLanguage);
+        ChromeDriver CreateDriverInner()
+        {
+            state.Options.AddArgument("--lang=" + configuration.AcceptLanguage);
 
-                // Disabling the Chrome sandbox can speed things up a bit, so recommended when you get a lot of timeouts
-                // during parallel execution:
-                // https://stackoverflow.com/questions/22322596/selenium-error-the-http-request-to-the-remote-webdriver-timed-out-after-60-sec
-                // However, this makes the executing machine vulnerable to browser-based attacks so it should only be
-                // used with trusted code (i.e. our own).
-                options.AddArgument("no-sandbox");
+            state.Options.SetLoggingPreference(LogType.Browser, LogLevel.Info);
 
-                if (configuration.Headless) options.AddArgument("headless");
+            // Disabling the Chrome sandbox can speed things up a bit, so recommended when you get a lot of timeouts
+            // during parallel execution:
+            // https://stackoverflow.com/questions/22322596/selenium-error-the-http-request-to-the-remote-webdriver-timed-out-after-60-sec
+            // However, this makes the executing machine vulnerable to browser-based attacks so it should only be used
+            // with trusted code (like our own).
+            state.Options.AddArgument("no-sandbox");
 
-                configuration.BrowserOptionsConfigurator?.Invoke(options);
+            // Linux-specific setting, may be necessary for running in containers, see
+            // https://developers.google.com/web/tools/puppeteer/troubleshooting#tips
+            state.Options.AddArgument("disable-dev-shm-usage");
 
-                return new ChromeDriver(ChromeDriverService.CreateDefaultService(), options, pageLoadTimeout).SetCommonTimeouts(pageLoadTimeout);
-            },
-            new ChromeConfig());
+            if (configuration.Headless) state.Options.AddArgument("headless");
+
+            configuration.BrowserOptionsConfigurator?.Invoke(state.Options);
+
+            state.Service ??= ChromeDriverService.CreateDefaultService();
+            state.Service.WhitelistedIPAddresses += "::ffff:127.0.0.1"; // By default localhost is only allowed in IPv4.
+            if (state.Service.HostName == "localhost") state.Service.HostName = "127.0.0.1"; // Helps with misconfigured hosts.
+
+            return new ChromeDriver(state.Service, state.Options, pageLoadTimeout).SetCommonTimeouts(pageLoadTimeout);
+        }
+
+        if (Environment.GetEnvironmentVariable("CHROMEWEBDRIVER") is { } driverPath &&
+            Directory.Exists(driverPath))
+        {
+            state.Service = ChromeDriverService.CreateDefaultService(driverPath);
+            return CreateDriverInner();
```

Comment:
I would prefer to move the `config` (`state`) variable into `CreateDriverInner()` and invoke the latter with a new parameter `service` which will be set to `ChromeDriverService.CreateDefaultService(driverPath)` here, and ... see my PR: https://github.com/Lombiq/UI-Testing-Toolbox/pull/167.